File: backend/app/schemas/kg_completion.py
from pydantic import BaseModel, Field


class KgCompletionModelMetric(BaseModel):
    """知识补全模型评估指标。"""

    model: str
    mrr: float = Field(..., ge=0.0, le=1.0)
    hits_at_1: float = Field(..., ge=0.0, le=1.0)
    hits_at_3: float = Field(..., ge=0.0, le=1.0)
    hits_at_10: float = Field(..., ge=0.0, le=1.0)
    train_seconds: int


class KgCompletionCurvePoint(BaseModel):
    """训练曲线中的单个轮次指标点。"""

    epoch: int
    transe: float
    complex: float
    rotate: float


class KgCompletionDataset(BaseModel):
    """知识补全实验使用的数据集统计。"""

    entity_count: int
    relation_count: int
    triple_count: int
    train_count: int
    valid_count: int
    test_count: int
    entity_types: dict[str, int]
    relation_types: dict[str, int]


class KgCompletionResponse(BaseModel):
    """知识补全实验总览响应。"""

    dataset: KgCompletionDataset
    model_metrics: list[KgCompletionModelMetric]
    mrr_curve: list[KgCompletionCurvePoint]
    hits_at_10_curve: list[KgCompletionCurvePoint]
    loss_curve: list[KgCompletionCurvePoint]
    conclusion: str


class KgCompletionPredictRequest(BaseModel):
    """Top-K 尾实体预测请求。"""

    head: str = Field(..., min_length=1)
    relation: str = Field(..., min_length=1)
    top_k: int = Field(default=3, ge=1, le=10)


class KgCompletionPrediction(BaseModel):
    """Top-K 补全候选结果。"""

    rank: int
    tail: str
    tail_name: str
    tail_type: str
    score: float = Field(..., ge=0.0, le=1.0)
    reason: str


class KgCompletionPredictResponse(BaseModel):
    """Top-K 尾实体预测响应。"""

    head: str
    relation: str
    predictions: list[KgCompletionPrediction]

File: frontend/src/api/kgCompletion.ts
import { apiClient } from "./client";
import type { ApiResponse } from "./mock";

export interface KgCompletionModelMetric {
  model: string;
  mrr: number;
  hits_at_1: number;
  hits_at_3: number;
  hits_at_10: number;
  train_seconds: number;
}

export interface KgCompletionCurvePoint {
  epoch: number;
  transe: number;
  complex: number;
  rotate: number;
}

export interface KgCompletionDataset {
  entity_count: number;
  relation_count: number;
  triple_count: number;
  train_count: number;
  valid_count: number;
  test_count: number;
  entity_types: Record<string, number>;
  relation_types: Record<string, number>;
}

export interface KgCompletionResponse {
  dataset: KgCompletionDataset;
  model_metrics: KgCompletionModelMetric[];
  mrr_curve: KgCompletionCurvePoint[];
  hits_at_10_curve: KgCompletionCurvePoint[];
  loss_curve: KgCompletionCurvePoint[];
  conclusion: string;
}

export interface KgCompletionPrediction {
  rank: number;
  tail: string;
  tail_name: string;
  tail_type: string;
  score: number;
  reason: string;
}

export interface KgCompletionPredictResponse {
  head: string;
  relation: string;
  predictions: KgCompletionPrediction[];
}

export async function fetchKgCompletionSummary(): Promise<KgCompletionResponse> {
  const response = await apiClient.get<ApiResponse<KgCompletionResponse>>("/api/experiments/kg-completion");
  return response.data.data;
}

export async function predictKgCompletion(
  head: string,
  relation: string,
  topK: number,
): Promise<KgCompletionPredictResponse> {
  const response = await apiClient.post<ApiResponse<KgCompletionPredictResponse>>(
    "/api/experiments/kg-completion/predict",
    {
      head,
      relation,
      top_k: topK,
    },
  );
  return response.data.data;
}
